### DISCUSSION TITLE
AttributeError: 'EdgeStorage' object has no attribute 'edge_index'

### DISCUSSION THREAD

**Discussion by gohinv** [pyg-team/pytorch_geometric#7824] [Q&A] [answered]
I am currently trying to create a link prediction model for a heterogeneous network with **2 node types and 3 edge types.** I have so far been able to preprocess and split my data into train, val, and test. Now I am trying to use NeighborLoader to create mini batches. Here is the line of code causing an error:

```python
train_loader = NeighborLoader(
    train_data,
    num_neighbors={key: [10] * 2 for key  in train_data.edge_types},
    batch_size=128,
    input_nodes=None,
)
```

resulting in:

```
KeyError                                  Traceback (most recent call last)
[/usr/local/lib/python3.10/dist-packages/torch_geometric/data/storage.py](https://localhost:8080/#) in __getattr__(self, key)
     78         try:
---> 79             return self[key]
     80         except KeyError:


[/usr/local/lib/python3.10/dist-packages/torch_geometric/data/storage.py](https://localhost:8080/#) in __getitem__(self, key)
    103     def __getitem__(self, key: str) -> Any:
--> 104         return self._mapping[key]
    105 

KeyError: 'edge_index'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
[<ipython-input-6-338c6780e0b4>](https://localhost:8080/#) in <cell line: 1>()
----> 1 train_loader = NeighborLoader(
      2     train_data,
      3     num_neighbors={key: [10] * 2 for key  in train_data.edge_types},
      4     batch_size=128,
      5     input_nodes=None,`
```

which traces back to:

```
[/usr/local/lib/python3.10/dist-packages/torch_geometric/data/storage.py](https://localhost:8080/#) in __getattr__(self, key)
     79             return self[key]
     80         except KeyError:
---> 81             raise AttributeError(
     82                 f"'{self.__class__.__name__}' object has no attribute '{key}'")
     83 

AttributeError: 'EdgeStorage' object has no attribute 'edge_index'
```

This is strange since I have verified that my dataset contains the `edge_index` Tensor for all three edge types. Thank you for the help.

**Comment by sauravkumr** [comment]
nah

**Comment by rusty1s** [comment, marked as answer]
Can you specify the `input_nodes`? I think this should crash prior already since you need to specify at least the `node_type` in `input_nodes` if you are operating on a heterogeneous graph.

**Reply by gohinv** [reply]
Thanks for the reply! I actually ended up using a LinkNeighborLoader instead. The error turned out that somehow my reverse edge data stores weren't being constructed properly for two of my edge types. I had to manually assign their attributes like edge_index, edge_label etc from the forward edges. It seemed to work otherwise, though.

**Reply by mdzamani** [reply]
Can you explain more about what you did? I encountered exactly the same problem.
Thanks.
